### DISCUSSION TITLE
Best Transfer Option for Large Deformation Transfer

### DISCUSSION THREAD

**Discussion by maxnezdyur** [idaholab/moose#22119] [Q&A General] [answered]
The images below present the problem I am having with the transfer. When there is a large deformation that transfer messes up. I tried a few different transfers, but I can't seem to get them to work. I am trying to transfer just a "1" from the solid onto the background mesh. This is to track the solid location within the fluid for an immersed FSI system I am trying to create. I made sure to use ` displaced_source_mesh = true`
![image](https://user-images.githubusercontent.com/22859100/190705225-a268b29b-c16d-4dbc-a8a4-8f3014a04d12.png)
![image](https://user-images.githubusercontent.com/22859100/190705253-65b5601f-f695-4dd5-8346-309a7963dbfc.png)

**Comment by GiudGiud** [comment]
Have you tried to set use_displaced_mesh to true in the transfer?

**Reply by maxnezdyur** [reply]
Yes, that didn't change anything. Also to keep in mind, the parent app doesn't have any displacements just the sub-app, so I don't think `use_displaced_mesh` would do anything anyways. I also tried this by switching the parent and sub app and tried `use_displaced_mesh ` and that didn't help either.

**Reply by GiudGiud** [reply]
I ll have to dig into it to check if we actually support subapp displacements in transfers.

Which transfer are you using? There are lots of options and some might support things better

**Reply by maxnezdyur** [reply]
I used MultiAppShapeEvaluationTransfer for the pictures above. I wanted to use MultiAppProjectionTransfer specifically, but for that one, the "indicator" never moved in the background mesh.

**Reply by GiudGiud** [reply]
What do you mean by the indicator?

Can you try GeometricInterpolation and NearestNode?

**Reply by maxnezdyur** [reply]
The indicator is the variable that I am transferring from the solid to the background mesh. Both of the options you describe "work" but I do not want the background mesh that is not near the solid to have the indicator value of 1. Both of the options you gave would make the entire background mesh equal to 1, if I didn't use my "hacky" method. This "hacky" method makes it so the boundary of the solid has value 0 so the nearest node for a background node that is not inside the solid is on the boundary. Below shows it works, but this method has faults because of the "hacky" method.
![image](https://user-images.githubusercontent.com/22859100/190714605-1b7ffe3e-019a-49da-af44-276d3f8283b8.png)

**Reply by GiudGiud** [reply]
I think you'd need block restriction to make it work without the accuracy issue
very few transfers have it, but you can try the ones in this PR which are yet to be merged
https://github.com/idaholab/moose/pull/17417

**Reply by maxnezdyur** [reply]
I am a bit confused. The background mesh is all 1 block. I need to use the transfer to determine which element gets changed to a new block to model the immersed solid. I can't block restrict the transfer because I do not which element is part of the solid or part of the real fluid before I transfer.

**Reply by lindsayad** [reply]
Is it possible for you to share a branch with your inputs (and any necessary code to run the inputs)?

**Reply by maxnezdyur** [reply]
I’ll create a minimal working example and send that over. I would have to share two branches and the current problem is solving the INS equations and that would be too long to run to debug anyway. Thanks for looking into this!

**Reply by hugary1995** [reply]
While @lindsayad looks into the issue with transfers involving displaced mesh, one workaround for the time being is to run both apps on the undisplaced mesh, and use the displacements to advect the indicator function.

**Comment by maxnezdyur** [comment]
Next time I will just share a branch, I didn't know I couldn't paste in a text file. In the parent file, I commented out a transfer that works with the method I described above but not perfectly. 

Parent File
```
[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = -1.0
    xmax = 3.0
    ymin = 0.0
    ymax = 1.0
    nx = 1000
    ny = 250
    elem_type = QUAD4
  []
[]
[Problem]
kernel_coverage_check = false
material_coverage_check = false
skip_nl_system_check = false
[]


[AuxVariables]
  [indicator]
    order = FIRST
    family = LAGRANGE
    initial_condition = 0.0
  []
[]

[Variables]
[dummy]

[]
[]

[Materials]
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'
  # Run for 100+ timesteps to reach stey state.

  dt = 1e-2
  end_time = 1
  dtmin = 1.0e-7
  # Note: -snes_ksp_ew seems to le to more nonlinear iterations, which isn't ideal
  # when compute_jacobian() is so expensive for this problem.
  petsc_options = '-snes_converged_reason -ksp_converged_reason'

  # # Direct solver
  petsc_options_iname = '-pc_type -pc_factor_shift_type -pc_factor_mat_solver_package'
  petsc_options_value = 'lu NONZERO superlu_dist'

  # petsc_options_iname = '-pc_type -pc_hypre_type'
  # petsc_options_value = 'hypre boomeramg'
  # residual_and_jacobian_together = true
  line_search = 'bt'
  nl_rel_tol = 1e-6
  nl_abs_tol = 2e-7
  nl_max_its = 5
  l_max_its = 25

  automatic_scaling = true
  off_diagonals_in_auto_scaling = true
  compute_scaling_once = false
  # scaling_group_variables = 'velocity p'
  [TimeIntegrator]
    type = BDF2
  []

[]

[Outputs]
  # [exo]
  #   type = Exodus
  # []
  exodus = true
  print_linear_residuals = false
[]

[MultiApps]
  [solid_domain]
    type = TransientMultiApp
    execute_on = "INITIAL TIMESTEP_END"
    # positions = '0 0 0'
    input_files = solid_multiapp.i
    use_displaced_mesh = true
    # sub_cycling = true
    catch_up = true
    max_catch_up_steps = 5
    keep_solution_during_restore = true
  []
[]

[Transfers]

  # [push_indicator]
  #   type = MultiAppGeometricInterpolationTransfer
  #   # Transfer from the sub-app from this app
  #   from_multi_app =  solid_domain
  #   # The name of the variable in this app
  #   source_variable = solid_indicator
  #   # The name of the auxiliary variable in the sub-app
  #   variable = indicator
  #   displaced_source_mesh = true
  #   use_displaced_mesh = true
  #   num_points = 1
  # []
  [push_indicator]
    type = MultiAppShapeEvaluationTransfer
    # Transfer from the sub-app from this app
    from_multi_app =  solid_domain
    # The name of the variable in this app
    source_variable = solid_indicator
    # The name of the auxiliary variable in the sub-app
    variable = indicator
    displaced_source_mesh = true
    use_displaced_mesh = true
  []
[]

```
Child below
```
beta = 0.25
gamma = 0.5

[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    ymin = 0
    xmax = 0.2
    ymax = 0.5
    nx = 50
    ny = 150
    elem_type = QUAD4
  []
[]
[Variables]
  [disp_x]
    order = FIRST
  []
  [disp_y]
    order = FIRST
  []
[]

[AuxVariables] # variables that are calculated for output

  [solid_indicator]
    order = FIRST
    family = LAGRANGE
    [AuxKernel]
      type = ConstantAux
      variable = solid_indicator
      value = 0.0
      boundary = 'left right top'
      execute_on = 'INITIAL TIMESTEP_END'
    []
    initial_condition = 1.0
  []
[]

[Modules/TensorMechanics/DynamicMaster]
  [all]
    # displacements = 'disp_x disp_y'
    add_variables = true
    # new_system = true
    incremental = true
    strain = FINITE
    decomposition_method = EigenSolution
    # hht_alpha = 0.25
  []
[]

[Materials]
  [elastic_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 10000.0
    poissons_ratio = 0.3
    use_displaced_mesh = true
  []
  [stress]
    type = ComputeFiniteStrainElasticStress
  []
  [density]
    type = GenericConstantMaterial
    prop_names = 'density'
    prop_values = '1'
    use_displaced_mesh = true
  []
  [constant_stress]
    type = GenericConstantRankTwoTensor
    tensor_values = '100'
    tensor_name = test_tensor
  []
[]

[BCs]
  [hold_x]
    type = DirichletBC
    boundary = bottom
    variable = disp_x
    value = 0
    use_displaced_mesh = true
    save_in = div_stess_x
  []
  [hold_y]
    type = DirichletBC
    boundary = bottom
    variable = disp_y
    value = 0
    use_displaced_mesh = true
    save_in = div_stess_y
  []
  [Pressure]
    [push_left]
      boundary = left
      factor = 100
    []
  []
[]

[Preconditioning]
  [SMP]
    type = SMP
    full = true
  []
[]
[Executioner]
  type = Transient
  end_time = 10
  dt = 1e-2
  solve_type = 'NEWTON'
  petsc_options = '-snes_converged_reason -ksp_converged_reason -snes_ksp_ew'
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_type -pc_factor_shift_type -pc_factor_shift_amount'
  petsc_options_value = 'lu       superlu_dist                  NONZERO               1e-15'
  nl_max_its = 40
  l_max_its = 15
  line_search = 'none'
  nl_abs_tol = 1e-5
  nl_rel_tol = 1e-4
  automatic_scaling = true
  [TimeIntegrator]
    type = NewmarkBeta
    beta = ${beta}
    gamma = ${gamma}
  []
[]

[Outputs]
  exodus = true
  print_linear_residuals = false
  print_linear_converged_reason = false
[]

# [MultiApps]
#   [fluid_domain]
#     type = TransientMultiApp
#     execute_on = "INITIAL TIMESTEP_END"
#     positions = '0 0 0'
#     input_files = real_fluid_multiapp.i
#     use_displaced_mesh = true
#   []
# []

# [Transfers]
#   [push_indicator]
#     type = MultiAppMeshFunctionTransfer
#     # Transfer to the sub-app from this app
#     to_multi_app = fluid_domain
#     # The name of the variable in this app
#     source_variable = solid_indicator
#     # The name of the auxiliary variable in the sub-app
#     variable = indicator
#     displaced_source_mesh = true
#     displaced_target_mesh = true
#     bbox_factor = 0.5
#     use_displaced_mesh = true
#   []
# []
```

**Comment by GiudGiud** [comment, marked as answer]
Any update on this?

**Reply by lindsayad** [reply]
I have not yet investigated

**Reply by lindsayad** [reply]
Ok, jotting some notes down here:

- libMesh node ID 89365 (exodus node ID 89366) is an example of a to-node getting a 0 value when it shouldn't
- This corresponds to the 4833 point that we want to query the from-mesh for (e.g. index 4832 in the incoming/outgoing points vector)
- When asking the mesh function to find the from-element containing that to-node we are getting back a null result, indicating that it thinks the point is out of the mesh
- When we get to the active nodes in `TreeNode`, we do get `bounds_point` but we do not get any elements that `contains_point`

I don't really know why this would be a hard case ... these are linear elements so I would expect inverse mapping from the physical space to the reference space to essentially be exact.

@roystgnr do you have any thoughts on what might be the issue here? Pasted below is a figure. The bottom is the sub-application field that we are transferring to the main application on top. You can see that we are getting these "teeth" in the transfer with the blue gaps in the teeth essentially indicating a failed transfer
<img width="739" alt="Screen Shot 2022-10-28 at 12 35 44 AM" src="https://user-images.githubusercontent.com/10248304/198531071-2a1ddc2f-fbac-4d66-b7d6-89bb7b149c19.PNG">

**Reply by lindsayad** [reply]
There is definitely some undesirable stuff happening here. More than enough to create an issue. Any further investigation will be discussed on the issue: #22534
